### PR TITLE
WE-698: Confirm operation meets rules completeness

### PR DESCRIPTION
**Pull request by ajaxscape**
I've nested the confirmRules router tests so the test report is nested, please let me know if you think it's a good idea.

### CODE REVIEW

File: src/controllers/confirmRules.controller.js [hunk 1]
```diff
@@ -6,11 +6,33 @@ const CookieService = require('../services/cookie.service')
 const LoggingService = require('../services/logging.service')
 const ConfirmRules = require('../models/confirmRules.model')
 
+function timeout (ms) {
```
Comment (duncannixon) [inline src/controllers/confirmRules.controller.js]: I have managed to reproduce the issue but I'd prefer not to have to use this timeout if possible. Please can you work with the Dynamics guys as they put some tracing on to see if we can get to the bottom of why the completion status doesn't seem to have been saved quickly enough. They should have some time to do this today apparently.

If we end up having to leave this in then a comment should be be added before the timeout() and untilComplete() methods to explain the reason for their existence (in addition to the comment you have already provided lower down at line 62).

File: src/controllers/confirmRules.controller.js [hunk 2]
```diff
@@ -24,19 +46,23 @@ module.exports = class ConfirmRulesController extends BaseController {
...
-      })
+        const complete = await this.isComplete(request)
+        if (complete) {
+          return reply.redirect(Constants.Routes.TASK_LIST.path)
+        }
 
-      try {
-        await confirmRules.save(authToken)
-        return reply.redirect(Constants.Routes.TASK_LIST.path)
+        const confirmRules = new ConfirmRules({
```
Comment (duncannixon) [inline src/controllers/confirmRules.controller.js]: I'd prefer to use non-ES6 syntax here to make it more maintainable by less experienced Javascript developers. 

In addition we are missing some properties here, we need applicationId and complete to be specified as well in the call to the constructor.

So it should be:

        const confirmRules = new ConfirmRules({
          applicationId: applicationId,
          applicationLineId: applicationLineId,
          complete: false
        })

Reason being that if we were using Typescript, for example, it wouldn't compile, and given our absence of compile-time type checking, I think that the interface is clearer and the code is more maintainable if we pass in all the properties, even if their values are undefined.

File: src/models/confirmRules.model.js [hunk 3]
```diff
@@ -6,10 +6,32 @@ const BaseModel = require('./base.model')
 const LoggingService = require('../services/logging.service')
 const ApplicationLine = require('./applicationLine.model')
 
-module.exports = class Site extends BaseModel {
+module.exports = class ConfirmRules extends BaseModel {
   constructor (data) {
     super()
-    this.applicationLineId = data.applicationLineId
+    // Merge the data with this instance
+    Object.assign(this, data)
```
Comment (duncannixon) [inline src/models/confirmRules.model.js]: As discussed I think we should keep using the non-ES6 syntax to set the instance properties in order to aid less experienced Javascript developers and to make it clear what the properties of the class are.

i.e.
The existing method of doing:
this.something = obj.something

File: src/controllers/confirmRules.controller.js (hunk 2 above)
Comment (ajaxscape) [reply src/controllers/confirmRules.controller.js]: Changed as requested.

File: src/controllers/confirmRules.controller.js (hunk 1 above)
Comment (ajaxscape) [reply src/controllers/confirmRules.controller.js]: I have moved this code into the confirmRules model and added comments as requested.

**Pull request merged**